src/coreclr/src/gc/gc.cpp

```diff
@@ -3312,24 +3382,345 @@ sorted_table::clear()
+            {
+                dprintf (1, ("Observed %Id free units in multiple blocks(%Id), coalescing",
+                    (size_t)num_contiguous_free_units, (size_t)last_num_free_units));
+                make_free_block (current_free_index_start, num_contiguous_free_units);
+            }
+            current_free_index_start = 0;
+            num_contiguous_free_units = 0;
+        }
+
+        current_index += current_num_units;
+    }
+
+    // If at this point current_free_index_start is not 0, it means we had
+    // free units at the end. Simply adjust used to be where the last free 
+    // block starts.
+    if (current_free_index_start != 0)
+    {
+        global_region_used = region_address_of (current_free_index_start);
+        region_map_end = current_free_index_start;
+    }
+
+    uint8_t* alloc = allocate_end_uh (num_units);
+
+    if (alloc)
+    {
+        print_map ("alloc: found at the end");
+    }
+    else
+    {
+        dprintf (1, ("couldn't find memory at the end! only %Id bytes left", (global_region_end - global_region_used)));
+    }
+
+    return alloc;
+}
+
+// ETW TODO: need to fire create seg events for these methods.
+// FIRE_EVENT(GCCreateSegment_V1
+bool region_allocator::allocate_region (size_t size, uint8_t** start, uint8_t** end)
+{
+    uint32_t* map_start = region_map_start;
+    size_t alignment = region_alignment;
+    size_t alloc_size = align_region_up (size);
+
+    uint32_t num_units = (uint32_t)(alloc_size / alignment);
+    bool ret = false;
+    uint8_t* alloc = NULL;
+    dprintf (1, ("----GET %d-----", num_units));
+
+    alloc = allocate (num_units);
+    *start = alloc;
+    *end = alloc + alloc_size;
+    ret = (alloc != NULL);
+
+    return ret;
+}
+
+bool region_allocator::allocate_basic_region (uint8_t** start, uint8_t** end)
+{
+    return allocate_region (region_alignment, start, end);
+}
+
+// Large regions are 8x basic region sizes by default. If you need a larger region than that,
+// call allocate_region with the size.
+bool region_allocator::allocate_large_region (uint8_t** start, uint8_t** end)
+{
+    return allocate_region (large_region_alignment, start, end);
+}
+
+void region_allocator::delete_region (uint8_t* start)
+{
+    assert (is_region_aligned (start));
+
+    print_map ("before delete");
+
+    uint32_t* current_index = region_map_index_of (start);
+    uint32_t current_val = *current_index;
+    assert (!is_unit_memory_free (current_val));
+
+    dprintf (1, ("----DEL %d-----", (current_index - region_map_start)));
+
+    make_free_block (current_index, current_val);
+
+    print_map ("after delete");
+}
+#endif //USE_REGIONS
+
 inline
 uint8_t* align_on_segment (uint8_t* add)
 {
-    return (uint8_t*)((size_t)(add + (gc_heap::min_segment_size - 1)) & ~(gc_heap::min_segment_size - 1));
+    return (uint8_t*)((size_t)(add + (((size_t)1 << gc_heap::min_segment_size_shr) - 1)) & ~(((size_t)1 << gc_heap::min_segment_size_shr) - 1));
```

Comment:
Is there a reason why this needed to change?

Reply:
yes because min_segment_size is not there when USE_REGIONS is defined.

Reply:
btw there's a bit more cleanup to do like I wanted to get rid of soh_segment_size too but I just haven't done that yet.

Reply:
Got it, thanks!